extensions/amp-story-auto-ads/0.1/algorithm-predetermined.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {StateProperty} from '../../amp-story/1.0/amp-story-store-service';
+
+/** @const {number} */
+const BEGINNING_OF_STORY_BUFFER = 3;
+
+/** @const {number} */
+const END_OF_STORY_BUFFER = 1;
+
+/** @const {number} */
+const MAX_ADS_PER_STORY = 4;
+
+/**
+ * Calculate the indices of where ads should be placed based
+ * on story length and the number of ads we want to show.
+ * @param {number} storyLength
+ * @param {number} numberOfAds
+ * @return {!Array<number>}
+ * @visibleForTesting
+ */
+export function getAdPositions(storyLength, numberOfAds) {
```

Comment:
Q: Is `numberOfAds` guaranteed to be non-zero?

Reply:
Nope, and I was not handling it correctly. Added early exit and test case. Good catch.